R/util_dfc.R

```diff
@@ -118,13 +118,178 @@ dfc.compare_df_values <- function(df1, df2, id_col, verbose = FALSE) {
+  ncols_match <- dim_comparison$ncols_comparison$current == dim_comparison$ncols_comparison$previous
+  
+  dim_comparison$nrows_comparison$match <- nrows_match
+  dim_comparison$ncols_comparison$match <- ncols_match
+  
+  # compare column names
+  colname_comparison <- dfc.setdiff_plus(names(current_df), names(previous_df))
+  colnames_match <- length(colname_comparison$only_in_first) == 0 & length(colname_comparison$only_in_second) == 0
+  
+  # compare indices
+  current_df$index <- dfc.get_concatenated_ids(current_df, index_cols)
+  previous_df$index <- dfc.get_concatenated_ids(previous_df, index_cols)
+  index_comparison <- dfc.setdiff_plus(current_df$index, previous_df$index)
+  indices_match <- length(index_comparison$only_in_first) == 0 & length(index_comparison$only_in_second) == 0
+  
+  # if everything else matches, and a values comparison was requested, go ahead and do a values comparison
+  if(nrows_match & ncols_match & colnames_match & indices_match & compare_values){
+    current_df <- current_df[order(current_df$index), ]
+    previous_df <- previous_df[order(previous_df$index), ]
+    values_comparison <- dfc.compare_df_values(current_df, previous_df, "index")
+    values_comparison$values_match <- all(values_comparison$col_summary$num_unmatched == 0)
+  } else{
+    values_comparison <- NA
+    side_by_side_df <- NA
+  }
+  comparison <- list(
+    "colname_comparison" = colname_comparison,
+    "index_comparison" = index_comparison,
+    "dim_comparison" = dim_comparison,
+    "values_comparison" = values_comparison
+  )
+  return(comparison)
 }
 
+compare_to_previous_summary <- function(comparison_to_previous){
+  # takes an object returned by compare_to_previous and summarizes deviations
+  # into readable df format (returns a data.frame)
+  n_cols_unmatched <- sum(
+    length(comparison_to_previous$colname_comparison$only_in_first),
+    length(comparison_to_previous$colname_comparison$only_in_second)
+  )
+  n_indices_only_in_current <- length(comparison_to_previous$index_comparison$only_in_first)
+  n_indices_only_in_previous <- length(comparison_to_previous$index_comparison$only_in_second)
+  nrows_diff <- comparison_to_previous$dim_comparison$nrows_comparison$current - 
+    comparison_to_previous$dim_comparison$nrows_comparison$previous
+  ncols_diff <- comparison_to_previous$dim_comparison$ncols_comparison$current - 
+    comparison_to_previous$dim_comparison$ncols_comparison$previous
+  
+  values_match <- NA
+  if(length(comparison_to_previous$values_comparison) > 1){
+    values_match <- all(comparison_to_previous$values_comparison$col_summary$num_unmatched == 0)
+  }
+  return(
+    data.frame(
+      n_cols_unmatched,
+      n_indices_only_in_current,
+      n_indices_only_in_previous,
+      nrows_diff,
+      ncols_diff,
+      values_match
+    )
+  )
+}
+
+dfc.get_granular_mismatches <- function(values_comparison, id_cols = "index"){
```

Comment:
This function is super awesome and worked great in my testing. But see my above comment about "id_cols" - I worry that it will be confusing to have all these nicely interoperable functions that unexpectedly make different assumptions about the id columns. I prefer requiring the user to supply one reliable ID column with the same name for both data frames, and identifying it in functions as the argument "id_col".

Reply:
Yeah, I agree with this point. My preference would be to work with vectors of id column names throughout because that's just more similar to what I do in the rest of my R-life.